File: app/src/main/java/com/esoxjem/movieguide/listing/MoviesListingFragment.java
package com.esoxjem.movieguide.listing;


import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.esoxjem.movieguide.BaseApplication;
import com.esoxjem.movieguide.Movie;
import com.esoxjem.movieguide.R;
import com.esoxjem.movieguide.listing.sorting.SortingDialogFragment;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class MoviesListingFragment extends Fragment implements IMoviesListingView
{
    @Inject
    IMoviesListingPresenter moviesPresenter;

    private RecyclerView.Adapter adapter;
    private List<Movie> movies = new ArrayList<>(20);
    private Callback callback;

    public MoviesListingFragment()
    {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context)
    {
        super.onAttach(context);
        callback = (Callback) context;
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        setRetainInstance(true);
        ((BaseApplication) getActivity().getApplication()).createListingComponent().inject(this);
        moviesPresenter.setView(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        View rootView = inflater.inflate(R.layout.fragment_movies, container, false);
        initLayoutReferences(rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState)
    {
        super.onViewCreated(view, savedInstanceState);
        moviesPresenter.displayMovies();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        switch (item.getItemId())
        {
            case R.id.action_sort:
                displaySortingOptions();
        }

        return super.onOptionsItemSelected(item);
    }

    private void displaySortingOptions()
    {
        DialogFragment sortingDialogFragment = SortingDialogFragment.newInstance(moviesPresenter);
        sortingDialogFragment.show(getFragmentManager(), "Select Quantity");
    }

    private void initLayoutReferences(View rootView)
    {
        RecyclerView moviesListing = (RecyclerView) rootView.findViewById(R.id.movies_listing);
        moviesListing.setHasFixedSize(true);

        int columns;
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE)
        {
            columns = 2;
        } else
        {
            columns = getResources().getInteger(R.integer.no_of_columns);
        }
        RecyclerView.LayoutManager layoutManager = new GridLayoutManager(getActivity(), columns);

        moviesListing.setLayoutManager(layoutManager);
        adapter = new MoviesListingAdapter(movies, this);
        moviesListing.setAdapter(adapter);
    }

    @Override
    public void showMovies(List<Movie> movies)
    {
        this.movies.clear();
        this.movies.addAll(movies);
        adapter.notifyDataSetChanged();
        callback.onMoviesLoaded(movies.get(0));
    }

    @Override
    public void loadingStarted()
    {
        Toast.makeText(getContext(), R.string.loading_movies, Toast.LENGTH_LONG).show();
    }

    @Override
    public void loadingFailed(String errorMessage)
    {
        Toast.makeText(getContext(), errorMessage, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onMovieClicked(Movie movie)
    {
        callback.onMovieClicked(movie);
    }

    @Override
    public void onDestroyView()
    {
        super.onDestroyView();
        moviesPresenter.destroy();
    }

    @Override
    public void onDetach()
    {
        callback = null;
        super.onDetach();
    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();
        ((BaseApplication) getActivity().getApplication()).releaseListingComponent();
    }

    public interface Callback
    {
        void onMoviesLoaded(Movie movie);

        void onMovieClicked(Movie movie);
    }
}

File: app/src/main/java/com/esoxjem/movieguide/movies/listing/MoviesListingFragment.kt
package com.esoxjem.movieguide.movies.listing


import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.os.Parcelable
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.esoxjem.movieguide.BaseApplication
import com.esoxjem.movieguide.R
import com.esoxjem.movieguide.movies.Constants
import com.esoxjem.movieguide.movies.entities.Movie
import com.esoxjem.movieguide.movies.sorting.SortingDialogFragment
import kotlinx.android.synthetic.main.fragment_movies.*
import java.util.ArrayList
import javax.inject.Inject

class MoviesListingFragment : Fragment(), MovieListingContract.View {
    @Inject
    lateinit var moviesPresenter: MovieListingContract.Presenter

    private lateinit var callback: Callback
    private lateinit var adapter: RecyclerView.Adapter<*>
    private var movies: MutableList<Movie> = ArrayList(20)

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callback = context as Callback
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        retainInstance = true
        (activity?.application as BaseApplication).createListingComponent().inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_movies, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initLayoutReferences()

        if (savedInstanceState != null) {
            movies = savedInstanceState.getParcelableArrayList(Constants.MOVIE)
            adapter.notifyDataSetChanged()
            movies_listing.visibility = View.VISIBLE
        } else {
            moviesPresenter.setView(this)
        }
    }

    private fun initLayoutReferences() {
        movies_listing.setHasFixedSize(true)
        movies_listing.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)

                if (recyclerView.canScrollVertically(1)) {
                    moviesPresenter.nextPage()
                }
            }
        })

        val columns: Int = if (resources.configuration.orientation
                == Configuration.ORIENTATION_LANDSCAPE) {
            2
        } else {
            resources.getInteger(R.integer.no_of_columns)
        }
        val layoutManager = GridLayoutManager(activity, columns)

        movies_listing.layoutManager = layoutManager
        adapter = MoviesListingAdapter(movies, this)
        movies_listing.adapter = adapter
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_sort -> {
                moviesPresenter.setView(this)
                displaySortingOptions()
            }
        }

        return super.onOptionsItemSelected(item)
    }

    private fun displaySortingOptions() {
        val sortingDialogFragment = SortingDialogFragment.newInstance()
        sortingDialogFragment.show(fragmentManager, "Select Quantity")
    }

    override fun showMovies(movies: List<Movie>) {
        this.movies.clear()
        this.movies.addAll(movies)
        movies_listing.visibility = View.VISIBLE
        adapter.notifyDataSetChanged()
        callback.onMoviesLoaded(movies[0])
    }

    override fun loadingStarted() {
        Snackbar.make(movies_listing, R.string.loading_movies, Snackbar.LENGTH_SHORT).show()
    }

    override fun loadingFailed(errorMessage: String?) {
        errorMessage?.let {
            Snackbar.make(movies_listing, errorMessage, Snackbar.LENGTH_INDEFINITE).show()
        }
    }

    override fun onMovieClicked(movie: Movie) {
        callback.onMovieClicked(movie)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        moviesPresenter.destroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(Constants.MOVIE, movies as ArrayList<out Parcelable>?)
    }


    interface Callback {
        fun onMoviesLoaded(movie: Movie)

        fun onMovieClicked(movie: Movie)
    }
}
